### DISCUSSION TITLE
External plugins

### DISCUSSION THREAD

**Discussion by MinekPo1** [DesModder/DesModder#500] [Ideas]
Currently, if I'm not mistaken, to create a new plugin, one must built the addon with it.

If a way to install external plugins was implemented, it would allow for easier development of small, niche plugins.

I'm thinking of using something like github releases?

**Comment by jared-hughes** [comment]
Right, all DesModder plugins are hardcoded in the DesModder source, so creating a new plugin requires a full development setup for the extension.

I agree, it would be nice to have an easy way to install external plugins (maybe by pasting URL to a github release, and there could be a browsable list of these). But I don't have a vision for how this could work well. It's a lot of work to create a system like this, a specialized variation on Tampermonkey.

Creating a plugin with DesModder (rather than Tampermonkey) currently has three main advantages:
1. on/off hooks in the DesModder menu
2. various helper functions (most get used by only one or two plugins)
3. (a big one) tooling support for UI components and typescript typings to integrate well with Desmos

The tooling support feels like the biggest benefit of developing plugins via DesModder, but it is hard to replicate for external plugins. I'd love to see how you might envision doing this externally.

I'd appreciate any ideas for how to approach external plugins.

**Reply by MinekPo1** [reply]
To be honest I have little to no experience with js/ts so it's hard for me to say. However maybe the web version of goosemod can serve as inspiration?

**Reply by jared-hughes** [reply]
Thanks for the input. It's a good idea to consider other modding platforms for inspiration. Some modding platforms I have experience with:

- (desktop apps) Minecraft Forge, OpenPlanet
- (electron app) BetterDiscord, Obsidian
- (web apps) Tampermonkey, Chrome Web Store

These all handle modding slightly differently, but something they all have in common is allowing for anyone to publish plugins/mods/userscripts/etc., and make these visible by search.
